src/modules/mc_att_control/mc_att_control_main.cpp

```diff
@@ -848,77 +850,26 @@ void
-	math::Matrix<3, 3> R_sp = q_sp.to_dcm();
-
-	/* get current rotation matrix from control state quaternions */
-	math::Quaternion q_att(_v_att.q[0], _v_att.q[1], _v_att.q[2], _v_att.q[3]);
-	math::Matrix<3, 3> R = q_att.to_dcm();
-
-	/* all input data is ready, run controller itself */
-
-	/* try to move thrust vector shortest way, because yaw response is slower than roll/pitch */
-	math::Vector<3> R_z(R(0, 2), R(1, 2), R(2, 2));
-	math::Vector<3> R_sp_z(R_sp(0, 2), R_sp(1, 2), R_sp(2, 2));
-
-	/* axis and sin(angle) of desired rotation */
-	math::Vector<3> e_R = R.transposed() * (R_z % R_sp_z);
-
-	/* calculate angle error */
-	float e_R_z_sin = e_R.length();
-	float e_R_z_cos = R_z * R_sp_z;
-
-	/* calculate weight for yaw control */
-	float yaw_w = R_sp(2, 2) * R_sp(2, 2);
+	using namespace matrix;
+	float yaw_w = .4f;
 
-	/* calculate rotation matrix after roll/pitch only rotation */
-	math::Matrix<3, 3> R_rp;
+	/* get estimated and desired vehicle attitude */
+	Quatf q(_v_att.q);
+	Quatf qd(_v_att_sp.q_d);
 
-	if (e_R_z_sin > 0.0f) {
-		/* get axis-angle representation */
-		float e_R_z_angle = atan2f(e_R_z_sin, e_R_z_cos);
-		math::Vector<3> e_R_z_axis = e_R / e_R_z_sin;
+	/* ensure quaternions are exactly normalized because acosf(1.00001) == NaN */
+	q.normalize();
+	qd.normalize();
 
-		e_R = e_R_z_axis * e_R_z_angle;
+	/* full quaternion attitude control, qe is rotation from q to qd */
+	Quatf qe = q.inversed() * qd;
 
-		/* cross product matrix for e_R_axis */
-		math::Matrix<3, 3> e_R_cp;
-		e_R_cp.zero();
-		e_R_cp(0, 1) = -e_R_z_axis(2);
-		e_R_cp(0, 2) = e_R_z_axis(1);
-		e_R_cp(1, 0) = e_R_z_axis(2);
-		e_R_cp(1, 2) = -e_R_z_axis(0);
-		e_R_cp(2, 0) = -e_R_z_axis(1);
-		e_R_cp(2, 1) = e_R_z_axis(0);
-
-		/* rotation matrix for roll/pitch only rotation */
-		R_rp = R * (_I + e_R_cp * e_R_z_sin + e_R_cp * e_R_cp * (1.0f - e_R_z_cos));
-
-	} else {
-		/* zero roll/pitch rotation */
-		R_rp = R;
-	}
-
-	/* R_rp and R_sp has the same Z axis, calculate yaw error */
-	math::Vector<3> R_sp_x(R_sp(0, 0), R_sp(1, 0), R_sp(2, 0));
-	math::Vector<3> R_rp_x(R_rp(0, 0), R_rp(1, 0), R_rp(2, 0));
-	e_R(2) = atan2f((R_rp_x % R_sp_x) * R_sp_z, R_rp_x * R_sp_x) * yaw_w;
-
-	if (e_R_z_cos < 0.0f) {
-		/* for large thrust vector rotations use another rotation method:
-		 * calculate angle and axis for R -> R_sp rotation directly */
-		math::Quaternion q_error;
-		q_error.from_dcm(R.transposed() * R_sp);
-		math::Vector<3> e_R_d = q_error(0) >= 0.0f ? q_error.imag()  * 2.0f : -q_error.imag() * 2.0f;
-
-		/* use fusion of Z axis based rotation and direct rotation */
-		float direct_w = e_R_z_cos * e_R_z_cos * yaw_w;
-		e_R = e_R * (1.0f - direct_w) + e_R_d * direct_w;
-	}
+	/* using sin(alpha/2) scaled rotation axis as attitude error (see quaternion definition by axis angle)
+	 * also taking care of the antipodal unit quaternion ambiguity */
+	Vector3f eq = 2.f * math::sign(qe(0)) * qe.imag();
+	math::Vector<3> e_R(eq.data());
```

Comment:
Can you do this without mixing math libraries?

Reply:
I got asked this before in reviews but in my opinion the matrix library is like the name suggests no general math library. I see the matrix library covering linear algebra and the math library more general math functions like sin, cos, sign, expo. That's why I also added all my useful mathematical helper functions here: https://github.com/PX4/Firmware/blob/13e64d00a8b8afa7cedd201ee85842de4feeff6a/src/lib/mathlib/math/Functions.hpp

Let me know if I got something wrong there.

Reply:
I'm fine with the existence of the math library, but I don't like that we have multiple types of Vector. Vector3f is from Matrix and math::Vector<3> is from the PX4 math library.

Reply:
I totally agree! I, like everyone else I talked to, favor the `matrix` library initiated by @jgoppert and that's why I contributed to it and look at `math::matrix` as deprecated.

Like in this pr I port every instance I come along to `matrix`. I also plan to get rid of the parameters:
https://github.com/PX4/Firmware/pull/8003/files#diff-81c4f50bdf837c4183b5d3cb81eb40eaR875
and outputs:
https://github.com/PX4/Firmware/pull/8003/files#diff-81c4f50bdf837c4183b5d3cb81eb40eaR872
in math::vector.

This is not to confuse with the call to `math::sign` which is a nonlinear scalar function that is defined in the `math` library because I felt it fits there best.

Reply:
I am bit confused as well. What is the reason for using math::Vector<3> instead of matrix?

Reply:
It's there to stay compatible with the rest while not replacing all data types.

But I see the consensus is that I should not only replace the attitude control functionality but totally switch the entire module to the `matrix` library.